### ISSUE TITLE
GMSH 4.1 / Elements with multiple physical tags / meshio ignores all but the first

### ISSUE THREAD

**Issue by hoelzlw** [nschloe/meshio#536]
I do not know, whether this is already mentioned somewhere, but I could not find it in the documentation.

I think, gmsh 4.0 is affected as well.

In gmsh one can assign an element to multiple physical groups, like in the example mesh file given below:
<details>

$MeshFormat
4.1 0 8
$EndMeshFormat
$PhysicalNames
7
1 3 "Mechanically Clamped"
1 4 "Mechanically Free"
1 5 "Electrically Clamped"
1 6 "Electrically Grounded"
1 7 "Electrically Free"
2 1 "Structural Mechanics"
2 2 "Piezoelectricity"
$EndPhysicalNames
$Entities
6 7 2 0
1 0 0 0 0 
2 2 0 0 0 
3 0 1 0 0 
4 2 1 0 0 
5 0 2 0 0 
6 2 2 0 0 
1 -9.999999994736442e-008 -1e-007 -1e-007 2.0000001 1e-007 1e-007 1 4 2 1 -2 
2 -9.999999994736442e-008 0.9999999000000001 -1e-007 2.0000001 1.0000001 1e-007 1 6 2 3 -4 
3 -9.999999994736442e-008 1.9999999 -1e-007 2.0000001 2.0000001 1e-007 2 4 5 2 5 -6 
4 1.9999999 -9.999999994736442e-008 -1e-007 2.0000001 1.0000001 1e-007 1 4 2 2 -4 
5 1.9999999 0.9999999000000001 -1e-007 2.0000001 2.0000001 1e-007 2 4 7 2 4 -6 
6 -1e-007 -9.999999994736442e-008 -1e-007 1e-007 1.0000001 1e-007 1 3 2 1 -3 
7 -1e-007 0.9999999000000001 -1e-007 1e-007 2.0000001 1e-007 2 3 7 2 3 -5 
1 -9.999999994736442e-008 -9.999999994736442e-008 -1e-007 2.0000001 1.0000001 1e-007 1 1 4 6 2 -4 -1 
2 -9.999999994736442e-008 0.9999999000000001 -1e-007 2.0000001 2.0000001 1e-007 2 1 2 4 7 3 -5 -2 
$EndEntities
$Nodes
15 13 1 13
0 1 0 1
1
0 0 0
0 2 0 1
2
2 0 0
0 3 0 1
3
0 1 0
0 4 0 1
4
2 1 0
0 5 0 1
5
0 2 0
0 6 0 1
6
2 2 0
1 1 0 1
7
1 0 0
1 2 0 1
8
1 1 0
1 3 0 1
9
1 2 0
1 4 0 0
1 5 0 0
1 6 0 0
1 7 0 0
2 1 0 2
10
11
1.5 0.5 0
0.5 0.5 0
2 2 0 2
12
13
1.5 1.5 0
0.5 1.5 0
$EndNodes
$Elements
9 26 1 26
1 1 1 2
1 1 7 
2 7 2 
1 2 1 2
3 3 8 
4 8 4 
1 3 1 2
5 5 9 
6 9 6 
1 4 1 1
7 2 4 
1 5 1 1
8 4 6 
1 6 1 1
9 1 3 
1 7 1 1
10 3 5 
2 1 2 8
11 3 11 1 
12 1 11 7 
13 2 10 4 
14 7 10 2 
15 8 11 3 
16 4 10 8 
17 8 10 7 
18 7 11 8 
2 2 2 8
19 5 13 3 
20 3 13 8 
21 4 12 6 
22 8 12 4 
23 9 13 5 
24 6 12 9 
25 9 12 8 
26 8 13 9 
$EndElements
</details>

This mesh denotes an example of a coupled physical problem, coupling structural mechanics and piezoelectricity.
The domain itself as well as parts of the boundary are assigned multiple physical groups, depending on their physics and boundary conditions in the mechanical and electrical domain.
For instance, the entire domain is marked as physical region 1 und part of it is additionally marked as physical region 2.

Now, meshio can read such a mesh and properly stores the different physical regions in the variable `field_data`:
```python
import meshio
infile = "/path/to/file.msh"
msh = meshio.read(infile)

msh.field_data
"""Output:
{'Electrically Clamped': array([5, 1]),
 'Electrically Free': array([7, 1]),
 'Electrically Grounded': array([6, 1]),
 'Mechanically Clamped': array([3, 1]),
 'Mechanically Free': array([4, 1]),
 'Piezoelectricity': array([2, 2]),
 'Structural Mechanics': array([1, 2])}
"""
```
But for the cells themselves, meshio throws away all but the first physical region.
```python
msh.cell_data['triangle']['gmsh:physical']
"""Output (piezoelectricity is gone)
array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
"""

msh.cell_data['line']['gmsh:physical']
"""Output (electrically clamped and free are gone)
array([4, 4, 6, 6, 4, 4, 4, 4, 3, 3])
"""
```

The reason is the code in https://github.com/nschloe/meshio/blob/8ed56d154ed896ac94cf925e8a41ea2987f3d747/meshio/_gmsh/msh4_1.py, function `_read_elements`.

The function _read_entities in line 101 properly constructs the variable `physical_tags` as a dictionary containing a list of boundary markers.
However, the function `_read_elements` in line 169 only ever accesses the first element of these lists (lines 220 and 228).

Right now, I don't know, how to solve this properly, so I suggest to add a warning in the function `_read_entities`, that the resulting mesh file will not reflect, what the user expected it to reflect.

Indeed, there is a workaround, by simply treating mixed boundaries differently and assigning them to a different pysical region for every possible combination:

<details>

$MeshFormat
4.1 0 8
$EndMeshFormat
$PhysicalNames
8
1 3 "Purely Mechanically Clamped"
1 4 "Purely Mechanically Free"
1 5 "Mixed Mechanically Clamped / Electrically Free"
1 6 "Mixed Mechanically Free / Electrically Clamped"
1 7 "Mixed Mechanically Free / Electrically Free"
1 8 "Purely Electrically Grounded"
2 1 "Pure Structural Mechanics"
2 2 "Mixed Structural Mechanics / Piezoelectricity"
$EndPhysicalNames
$Entities
6 7 2 0
1 0 0 0 0 
2 2 0 0 0 
3 0 1 0 0 
4 2 1 0 0 
5 0 2 0 0 
6 2 2 0 0 
1 -9.999999994736442e-008 -1e-007 -1e-007 2.0000001 1e-007 1e-007 1 4 2 1 -2 
2 -9.999999994736442e-008 0.9999999000000001 -1e-007 2.0000001 1.0000001 1e-007 1 8 2 3 -4 
3 -9.999999994736442e-008 1.9999999 -1e-007 2.0000001 2.0000001 1e-007 1 6 2 5 -6 
4 1.9999999 -9.999999994736442e-008 -1e-007 2.0000001 1.0000001 1e-007 1 4 2 2 -4 
5 1.9999999 0.9999999000000001 -1e-007 2.0000001 2.0000001 1e-007 1 7 2 4 -6 
6 -1e-007 -9.999999994736442e-008 -1e-007 1e-007 1.0000001 1e-007 1 3 2 1 -3 
7 -1e-007 0.9999999000000001 -1e-007 1e-007 2.0000001 1e-007 1 5 2 3 -5 
1 -9.999999994736442e-008 -9.999999994736442e-008 -1e-007 2.0000001 1.0000001 1e-007 1 1 4 6 2 -4 -1 
2 -9.999999994736442e-008 0.9999999000000001 -1e-007 2.0000001 2.0000001 1e-007 1 2 4 7 3 -5 -2 
$EndEntities
$Nodes
15 13 1 13
0 1 0 1
1
0 0 0
0 2 0 1
2
2 0 0
0 3 0 1
3
0 1 0
0 4 0 1
4
2 1 0
0 5 0 1
5
0 2 0
0 6 0 1
6
2 2 0
1 1 0 1
7
1 0 0
1 2 0 1
8
1 1 0
1 3 0 1
9
1 2 0
1 4 0 0
1 5 0 0
1 6 0 0
1 7 0 0
2 1 0 2
10
11
1.5 0.5 0
0.5 0.5 0
2 2 0 2
12
13
1.5 1.5 0
0.5 1.5 0
$EndNodes
$Elements
9 26 1 26
1 1 1 2
1 1 7 
2 7 2 
1 2 1 2
3 3 8 
4 8 4 
1 3 1 2
5 5 9 
6 9 6 
1 4 1 1
7 2 4 
1 5 1 1
8 4 6 
1 6 1 1
9 1 3 
1 7 1 1
10 3 5 
2 1 2 8
11 3 11 1 
12 1 11 7 
13 2 10 4 
14 7 10 2 
15 8 11 3 
16 4 10 8 
17 8 10 7 
18 7 11 8 
2 2 2 8
19 5 13 3 
20 3 13 8 
21 4 12 6 
22 8 12 4 
23 9 13 5 
24 6 12 9 
25 9 12 8 
26 8 13 9 
$EndElements
</details>

This leads to
```python
msh.field_data
"""Output
{'Mixed Mechanically Clamped / Electrically Free': array([5, 1]),
 'Mixed Mechanically Free / Electrically Clamped': array([6, 1]),
 'Mixed Mechanically Free / Electrically Free': array([7, 1]),
 'Mixed Structural Mechanics / Piezoelectricity': array([2, 2]),
 'Pure Structural Mechanics': array([1, 2]),
 'Purely Electrically Grounded': array([8, 1]),
 'Purely Mechanically Clamped': array([3, 1]),
 'Purely Mechanically Free': array([4, 1])}
"""

msh.cell_data['triangle']['gmsh:physical']
"""Output
array([1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2])
"""

msh.cell_data['line']['gmsh:physical']
"""Output
array([4, 4, 8, 8, 6, 6, 4, 7, 3, 5])
"""
```

This, however, can become very tedious for complicated geometries.

Best regards
Wolfgang

**Comment by MEMAndersen** [comment]
I just want to chime in and say that this is something I am also interested in finding a solution to.

**Comment by gdmcbain** [comment]
Isn't #698 the solution?

**Comment by MEMAndersen** [comment]
Since this issue was still open i thought it wasn't (i am not so familar with github). I will experiment with the package agian. Thanks for the heads up

**Comment by gdmcbain** [comment]
Sorry, yes, the intention of #698 was very much to fix issues just like this. I knew that there were a few outstanding but couldn't find them all. I'm away from the computer over the weekend but will verify the fix on Monday. Do let us know of failure or success in your case too.

**Comment by gdmcbain** [comment]
Now reading the original example of @hoelzlw, I get:
```
>>> mesh.cell_sets_dict['Mechanically Clamped']
{'triangle': array([0, 8], dtype=uint64)}
>>> mesh.cell_sets_dict['Mechanically Free']
{'line': array([0, 1, 4, 5, 7, 8], dtype=uint64)}
>>> mesh.cell_sets_dict['Electrically Clamped']
{'line': array([6, 7], dtype=uint64)}
>>> mesh.cell_sets_dict['Electrically Grounded']
{'line': array([2, 3], dtype=uint64)}
>>> mesh.cell_sets_dict['Electrically Free']
{'line': array([9], dtype=uint64)}
>>> mesh.cell_sets_dict['Structural Mechanics']
{}
>>> mesh.cell_sets_dict['Piezoelectricity']
{}
```
The boundaries look O. K., but something has clearly gone very wrong with the subdomains!  Especially since they do appear in the `.cell_sets` corresponding to the 'triangle' elements (the last two):
```
>>> meshio.cell_sets['Piezoelectricity'][-2:]
[array([], dtype=uint64), array([0, 1, 2, 3, 4, 5, 6, 7], dtype=uint64)]
>>> meshio.cell_sets['Structural Mechanics'][-2:]
[array([0, 1, 2, 3, 4, 5, 6, 7], dtype=uint64), array([], dtype=uint64)]
```

**Comment by gdmcbain** [comment]
The [_read_elements](https://github.com/nschloe/meshio/blob/8ed56d154ed896ac94cf925e8a41ea2987f3d747/meshio/_gmsh/msh4_1.py#L220-L228) cited above indeed ignored tags past the first, but that's because of the limitation of `Mesh.cell_data`; however, the newer code in the same function that was supposed to put all the tags into the `.cell_sets` attribute is buggy, appending a new array of cell indices to each item whenever a physical tag is processed instead of only for each entity block.

https://github.com/nschloe/meshio/blob/add5d452f853cca3f2cb4a9151b297dda03442ce/meshio/gmsh/_gmsh41.py#L193-L207

  This accidentally works whenever there is exactly one physical tag per block (which was the case in the examples used during development) but is wrong in general and in particular for the example given in this issue.

**Comment by gdmcbain** [comment]
It should use `cell_set[k]` (`k` being otherwise unused) rather than `cell_set.append`.  Or something like that—perhaps `cell-sets` should be initialized differently.  I'll fix this tomorrow if no-one beats me to it.

**Comment by hoelzlw** [comment]
@gdmcbain Off-Topic: Just a question for the record about the protocol here in this project:
Am I, the OP, supposed to close this issue once it's fixed, or will you people do this?

**Comment by gdmcbain** [comment]
I hadn't thought about that. Often a PR will contain a comment like 'fixes #536' and then GitHub will automatically close the issue when it's merged. I guess if the OP isn't satisfied they can reopen the issue or launch another precising the shortcoming?

**Comment by gdmcbain** [comment]
O. K., branch gdmcbain:536-msh41-multiple-tags in PR #756 seems to do the job:
```python
for k, v in mesh.cell_sets_dict.items():
    print(k, ':', next(map(list, v.values())))
```
gives
```
Mechanically Clamped : [8, 9]
Mechanically Free : [0, 1, 4, 5, 6, 7]
Electrically Clamped : [4, 5]
Electrically Grounded : [2, 3]
Electrically Free : [7, 9]
Structural Mechanics : [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
Piezoelectricity : [8, 9, 10, 11, 12, 13, 14, 15]
```
